File: src/rlm_research_planner/services/castle_planning.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Mapping

from rlm_research_planner.services.calculation import (
    apply_free_speedup_time,
    apply_guild_helps,
    apply_research_speed,
    free_speedup_seconds_for_vip,
)


CASTLE_RESOURCE_KEYS = (
    "food",
    "stone",
    "timber",
    "ore",
    "gold_hammer",
    "war_tome",
    "steel_cuffs",
    "soul_crystal",
    "mana_ore",
    "mana_crystal",
    "mana_steel",
)


def minimum_gems_for_amount(
    amount: int,
    packs: tuple[tuple[int, int], ...],
) -> int:
    """Return the cheapest gem cost for buying at least ``amount`` items."""

    required = max(0, int(amount))
    normalized = tuple(
        (max(1, int(quantity)), max(0, int(gems)))
        for quantity, gems in packs
        if int(quantity) > 0 and int(gems) >= 0
    )
    if required == 0:
        return 0
    if not normalized:
        raise ValueError("Gem-shop pack data is missing")
    maximum_quantity = max(quantity for quantity, _gems in normalized)
    limit = required + maximum_quantity - 1
    unreachable = 10**30
    costs = [unreachable] * (limit + 1)
    costs[0] = 0
    for owned in range(limit + 1):
        if costs[owned] == unreachable:
            continue
        for quantity, gems in normalized:
            next_amount = min(limit, owned + quantity)
            costs[next_amount] = min(costs[next_amount], costs[owned] + gems)
    return min(costs[required:])


@dataclass(frozen=True)
class BuildingLevelData:
    building_id: str
    level: int
    base_time_seconds: int
    costs: Mapping[str, int]
    requirements: tuple[tuple[str, int], ...]


@dataclass(frozen=True)
class BuildingData:
    id: str
    names: Mapping[str, str]
    max_level: int
    levels: Mapping[int, BuildingLevelData]

    def localized_name(self, locale: str) -> str:
        return (
            self.names.get(locale)
            or self.names.get(locale.split("-", 1)[0])
            or self.names.get("en-US")
            or self.id
        )


@dataclass(frozen=True)
class CastleManaStageData:
    stage: int
    base_time_seconds: int
    costs: Mapping[str, int]


@dataclass(frozen=True)
class CastlePlanStep:
    building_id: str
    level: int
    base_seconds: int
    adjusted_seconds: int
    costs: Mapping[str, int]
    mana_stage: int = 0


@dataclass(frozen=True)
class CastleBuildingSummary:
    building_id: str
    current_level: int
    target_level: int
    base_seconds: int
    adjusted_seconds: int
    costs: Mapping[str, int]


@dataclass(frozen=True)
class CastlePlanResult:
    current_castle_level: int
    target_castle_level: int
    current_mana_stage: int
    target_mana_stage: int
    target_building_id: str
    target_building_level: int
    effective_levels: Mapping[str, int]
    steps: tuple[CastlePlanStep, ...]
    buildings: tuple[CastleBuildingSummary, ...]
    total_base_seconds: int
    total_adjusted_seconds: int
    total_costs: Mapping[str, int]
    gem_costs: Mapping[str, int]
    total_gems: int
    issues: tuple[str, ...] = ()


class CastleCatalog:
    def __init__(
        self,
        buildings: Mapping[str, BuildingData],
        mana_stages: Mapping[int, CastleManaStageData] | None = None,
        gem_shop_packs: Mapping[str, tuple[tuple[int, int], ...]] | None = None,
    ) -> None:
        self.buildings = dict(buildings)
        self.mana_stages = dict(mana_stages or {})
        self.gem_shop_packs = dict(gem_shop_packs or {})

    @property
    def max_mana_stage(self) -> int:
        return max(self.mana_stages, default=0)

    @classmethod
    def load(cls, path: Path) -> "CastleCatalog":
        raw = json.loads(path.read_text(encoding="utf-8"))
        buildings: dict[str, BuildingData] = {}
        for source in raw.get("buildings", []):
            building_id = str(source["id"])
            levels: dict[int, BuildingLevelData] = {}
            for level_text, level_source in source.get("levels", {}).items():
                level = int(level_text)
                levels[level] = BuildingLevelData(
                    building_id=building_id,
                    level=level,
                    base_time_seconds=max(
                        0, int(level_source.get("base_time_seconds") or 0)
                    ),
                    costs={
                        key: max(0, int(level_source.get("costs", {}).get(key, 0)))
                        for key in CASTLE_RESOURCE_KEYS
                    },
                    requirements=tuple(
                        (
                            str(requirement["building_id"]),
                            max(1, int(requirement["level"])),
                        )
                        for requirement in level_source.get("requirements", [])
                    ),
                )
            buildings[building_id] = BuildingData(
                id=building_id,
                names={str(key): str(value) for key, value in source["names"].items()},
                max_level=max(1, int(source.get("max_level", 25))),
                levels=levels,
            )
        mana_stages: dict[int, CastleManaStageData] = {}
        mana_source = raw.get("castle_mana_progression", {})
        for stage_text, stage_source in mana_source.get("stages", {}).items():
            stage = int(stage_text)
            mana_stages[stage] = CastleManaStageData(
                stage=stage,
                base_time_seconds=max(
                    0, int(stage_source.get("base_time_seconds") or 0)
                ),
                costs={
                    key: max(0, int(stage_source.get("costs", {}).get(key, 0)))
                    for key in CASTLE_RESOURCE_KEYS
                },
            )
        gem_shop_packs = {
            str(key): tuple(
                (
                    max(1, int(pack.get("quantity") or 1)),
                    max(0, int(pack.get("gems") or 0)),
                )
                for pack in packs
            )
            for key, packs in raw.get("gem_shop_packs", {}).items()
        }
        if "castle" not in buildings:
            raise ValueError("Castle data is missing")
        return cls(buildings, mana_stages, gem_shop_packs)

    def minimum_levels_for_castle(self, castle_level: int) -> dict[str, int]:
        levels: dict[str, int] = {building_id: 0 for building_id in self.buildings}
        visiting: set[tuple[str, int]] = set()

        def require(building_id: str, target_level: int) -> None:
            building = self.buildings.get(building_id)
            if not building:
                return
            normalized = min(building.max_level, max(0, int(target_level)))
            if normalized <= levels[building_id]:
                return
            previous = levels[building_id]
            levels[building_id] = normalized
            for level in range(previous + 1, normalized + 1):
                key = (building_id, level)
                if key in visiting:
                    raise ValueError(
                        f"Cyclic building prerequisite detected at {building_id}:{level}"
                    )
                visiting.add(key)
                data = building.levels.get(level)
                if data:
                    for required_id, required_level in data.requirements:
                        require(required_id, required_level)
                visiting.remove(key)

        require("castle", castle_level)
        return levels

    def effective_levels(
        self, castle_level: int, saved_levels: Mapping[str, int] | None = None
    ) -> dict[str, int]:
        result = self.minimum_levels_for_castle(castle_level)
        for building_id, value in (saved_levels or {}).items():
            building = self.buildings.get(building_id)
            if not building:
                continue
            result[building_id] = max(
                result.get(building_id, 0),
                min(building.max_level, max(0, int(value))),
            )
        result["castle"] = max(result.get("castle", 0), int(castle_level))
        return result

    def gem_costs_for(
        self,
        costs: Mapping[str, int],
        owned_resources: Mapping[str, int] | None = None,
    ) -> dict[str, int]:
        available = owned_resources or {}
        result: dict[str, int] = {}
        for key, packs in self.gem_shop_packs.items():
            missing = max(
                0,
                int(costs.get(key, 0)) - max(0, int(available.get(key, 0))),
            )
            if missing:
                result[key] = minimum_gems_for_amount(missing, packs)
        return result

    def create_plan(
        self,
        *,
        castle_level: int,
        target_castle_level: int,
        current_mana_stage: int = 0,
        target_mana_stage: int = 0,
        saved_levels: Mapping[str, int] | None = None,
        construction_speed_percent: float = 0.0,
        vip_level: int = 1,
        guild_helps: int = 0,
        target_building_id: str = "castle",
        target_building_level: int | None = None,
        owned_resources: Mapping[str, int] | None = None,
    ) -> CastlePlanResult:
        castle = self.buildings["castle"]
        current = min(castle.max_level, max(1, int(castle_level)))
        selected_building = self.buildings.get(target_building_id)
        if selected_building is None:
            raise KeyError(target_building_id)
        target = min(castle.max_level, max(current, int(target_castle_level)))
        selected_target = (
            target
            if target_building_id == "castle"
            else min(
                selected_building.max_level,
                max(
                    self.effective_levels(current, saved_levels).get(
                        target_building_id, 0
                    ),
                    int(target_building_level or 0),
                ),
            )
        )
        if target_building_id != "castle":
            target = current
        current_mana = (
            min(self.max_mana_stage, max(0, int(current_mana_stage)))
            if current >= castle.max_level
            else 0
        )
        target_mana = (
            min(
                self.max_mana_stage,
                max(current_mana, int(target_mana_stage)),
            )
            if target >= castle.max_level
            else 0
        )
        effective = self.effective_levels(current, saved_levels)
        steps: list[CastlePlanStep] = []
        completed: set[tuple[str, int]] = set()
        visiting: set[tuple[str, int]] = set()
        issues: list[str] = []

        def add_building(building_id: str, target_level: int) -> None:
            building = self.buildings.get(building_id)
            if not building:
                issues.append(f"Unknown building: {building_id}")
                return
            normalized = min(building.max_level, max(0, int(target_level)))
            start = effective.get(building_id, 0) + 1
            for level in range(start, normalized + 1):
                key = (building_id, level)
                if key in completed:
                    continue
                if key in visiting:
                    raise ValueError(
                        f"Cyclic building prerequisite detected at {building_id}:{level}"
                    )
                visiting.add(key)
                data = building.levels.get(level)
                if not data:
                    issues.append(f"Missing building data: {building_id}:{level}")
                    visiting.remove(key)
                    continue
                for required_id, required_level in data.requirements:
                    add_building(required_id, required_level)
                adjusted = apply_research_speed(
                    data.base_time_seconds, construction_speed_percent
                )
                adjusted = apply_free_speedup_time(
                    adjusted, free_speedup_seconds_for_vip(vip_level)
                )
                adjusted = apply_guild_helps(adjusted, max(0, int(guild_helps)))
                steps.append(
                    CastlePlanStep(
                        building_id=building_id,
                        level=level,
                        base_seconds=data.base_time_seconds,
                        adjusted_seconds=adjusted,
                        costs=data.costs,
                    )
                )
                completed.add(key)
                visiting.remove(key)

        add_building(target_building_id, selected_target)
        for stage in range(
            current_mana + 1,
            target_mana + 1 if target_building_id == "castle" else current_mana + 1,
        ):
            data = self.mana_stages.get(stage)
            if data is None:
                issues.append(f"Missing Castle Mana stage data: {stage}")
                continue
            adjusted = apply_research_speed(
                data.base_time_seconds, construction_speed_percent
            )
            adjusted = apply_free_speedup_time(
                adjusted, free_speedup_seconds_for_vip(vip_level)
            )
            adjusted = apply_guild_helps(adjusted, max(0, int(guild_helps)))
            steps.append(
                CastlePlanStep(
                    building_id="castle",
                    level=castle.max_level,
                    base_seconds=data.base_time_seconds,
                    adjusted_seconds=adjusted,
                    costs=data.costs,
                    mana_stage=stage,
                )
            )
        summary_by_id: dict[str, dict[str, object]] = {}
        total_costs = {key: 0 for key in CASTLE_RESOURCE_KEYS}
        for step in steps:
            summary = summary_by_id.setdefault(
                step.building_id,
                {
                    "target_level": effective.get(step.building_id, 0),
                    "base_seconds": 0,
                    "adjusted_seconds": 0,
                    "costs": {key: 0 for key in CASTLE_RESOURCE_KEYS},
                },
            )
            summary["target_level"] = max(int(summary["target_level"]), step.level)
            summary["base_seconds"] = int(summary["base_seconds"]) + step.base_seconds
            summary["adjusted_seconds"] = (
                int(summary["adjusted_seconds"]) + step.adjusted_seconds
            )
            costs = summary["costs"]
            assert isinstance(costs, dict)
            for key in CASTLE_RESOURCE_KEYS:
                amount = int(step.costs.get(key, 0))
                costs[key] = int(costs[key]) + amount
                total_costs[key] += amount
        summaries = tuple(
            CastleBuildingSummary(
                building_id=building_id,
                current_level=effective.get(building_id, 0),
                target_level=int(summary["target_level"]),
                base_seconds=int(summary["base_seconds"]),
                adjusted_seconds=int(summary["adjusted_seconds"]),
                costs=dict(summary["costs"]),
            )
            for building_id, summary in summary_by_id.items()
        )
        gem_costs = self.gem_costs_for(total_costs, owned_resources)
        return CastlePlanResult(
            current_castle_level=current,
            target_castle_level=target,
            current_mana_stage=current_mana,
            target_mana_stage=(
                target_mana if target_building_id == "castle" else current_mana
            ),
            target_building_id=target_building_id,
            target_building_level=selected_target,
            effective_levels=effective,
            steps=tuple(steps),
            buildings=summaries,
            total_base_seconds=sum(step.base_seconds for step in steps),
            total_adjusted_seconds=sum(step.adjusted_seconds for step in steps),
            total_costs=total_costs,
            gem_costs=gem_costs,
            total_gems=sum(gem_costs.values()),
            issues=tuple(dict.fromkeys(issues)),
        )

File: tools/RLMResearchPlannerPWA/src/castle-planning.js
import { freeSecondsForVip, guildHelpCount } from "./state.js?v=0.1.1-b2";
import { loadJsonResource } from "./catalog.js?v=0.1.1-b2";

export const CASTLE_RESOURCE_KEYS = ["food", "stone", "timber", "ore", "gold_hammer", "war_tome", "steel_cuffs", "soul_crystal", "mana_ore", "mana_crystal", "mana_steel"];

function localText(values, locale) {
  return values?.[locale] || values?.[locale?.split("-")[0]] || values?.["en-US"] || Object.values(values || {})[0] || "";
}

export async function loadCastleCatalog(url = "./data/buildings/castle_catalog.json") {
  return normalizeCastleCatalog(await loadJsonResource(url, "城・施設データ"));
}

export function normalizeCastleCatalog(raw) {
  let languagePack = null;
  const buildings = new Map();
  for (const source of raw.buildings || []) {
    const levels = new Map();
    for (const [levelText, level] of Object.entries(source.levels || {})) {
      levels.set(Number(levelText), {
        level: Number(levelText),
        baseTimeSeconds: Math.max(0, Number(level.base_time_seconds) || 0),
        costs: Object.fromEntries(CASTLE_RESOURCE_KEYS.map((key) => [key, Math.max(0, Number(level.costs?.[key]) || 0)])),
        requirements: (level.requirements || []).map((item) => ({ buildingId: String(item.building_id), level: Math.max(1, Number(item.level) || 1) })),
      });
    }
    buildings.set(source.id, { id: source.id, names: { ...(source.names || {}) }, maxLevel: Number(source.max_level || 25), levels });
  }
  if (!buildings.has("castle")) throw new Error("城データがありません");
  const manaSource = raw.castle_mana_progression || {};
  const manaStages = new Map(Object.entries(manaSource.stages || {}).map(([stageText, source]) => {
    const stage = Number(stageText);
    return [stage, {
      stage,
      baseTimeSeconds: Math.max(0, Number(source.base_time_seconds) || 0),
      costs: Object.fromEntries(CASTLE_RESOURCE_KEYS.map((key) => [key, Math.max(0, Number(source.costs?.[key]) || 0)])),
    }];
  }));
  const gemShopPacks = Object.fromEntries(Object.entries(raw.gem_shop_packs || {}).map(([key, packs]) => [key, (packs || []).map((pack) => ({
    quantity: Math.max(1, Math.trunc(Number(pack.quantity) || 1)),
    gems: Math.max(0, Math.trunc(Number(pack.gems) || 0)),
  }))]));
  return {
    buildings,
    manaStages,
    maxManaStage: Math.max(0, ...manaStages.keys()),
    manaNames: { ...(manaSource.names || {}) },
    gemShopPacks,
    order: [...buildings.keys()],
    setLanguagePack(pack) { languagePack = pack || null; },
    sourceBuildingName(buildingId, locale) { const building = buildings.get(buildingId); return building ? localText(building.names, locale) : buildingId; },
    buildingName(buildingId, locale) { const building = buildings.get(buildingId); return languagePack?.sections?.buildings?.[buildingId] || (building ? localText(building.names, languagePack?.fallbackLocale || locale) : buildingId); },
    manaName(locale) { return localText(this.manaNames, locale) || this.buildingName("castle", locale); },
  };
}

export function minimumGemsForAmount(amount, packs) {
  const required = Math.max(0, Math.trunc(Number(amount) || 0));
  const normalized = (packs || []).map((pack) => ({
    quantity: Math.max(1, Math.trunc(Number(pack.quantity) || 1)),
    gems: Math.max(0, Math.trunc(Number(pack.gems) || 0)),
  })).filter((pack) => pack.quantity > 0);
  if (!required) return 0;
  if (!normalized.length) throw new Error("ジェムショップの購入単位がありません");
  const limit = required + Math.max(...normalized.map((pack) => pack.quantity)) - 1;
  const costs = Array(limit + 1).fill(Number.POSITIVE_INFINITY);
  costs[0] = 0;
  for (let owned = 0; owned <= limit; owned += 1) {
    if (!Number.isFinite(costs[owned])) continue;
    for (const pack of normalized) {
      const next = Math.min(limit, owned + pack.quantity);
      costs[next] = Math.min(costs[next], costs[owned] + pack.gems);
    }
  }
  return Math.min(...costs.slice(required));
}

export function gemCostsFor(catalog, costs, ownedResources = {}) {
  return Object.fromEntries(Object.entries(catalog.gemShopPacks || {}).flatMap(([key, packs]) => {
    const missing = Math.max(0, Number(costs?.[key] || 0) - Number(ownedResources?.[key] || 0));
    return missing > 0 ? [[key, minimumGemsForAmount(missing, packs)]] : [];
  }));
}

export function minimumBuildingLevels(catalog, castleLevel) {
  const levels = Object.fromEntries(catalog.order.map((id) => [id, 0]));
  const visiting = new Set();
  const requireBuilding = (buildingId, targetLevel) => {
    const building = catalog.buildings.get(buildingId);
    if (!building) return;
    const target = Math.min(building.maxLevel, Math.max(0, Math.trunc(Number(targetLevel) || 0)));
    if (target <= Number(levels[buildingId] || 0)) return;
    const previous = Number(levels[buildingId] || 0);
    levels[buildingId] = target;
    for (let level = previous + 1; level <= target; level += 1) {
      const key = `${buildingId}\0${level}`;
      if (visiting.has(key)) throw new Error(`施設の前提条件が循環しています: ${buildingId}:${level}`);
      visiting.add(key);
      for (const requirement of building.levels.get(level)?.requirements || []) requireBuilding(requirement.buildingId, requirement.level);
      visiting.delete(key);
    }
  };
  requireBuilding("castle", castleLevel);
  return levels;
}

function effectiveBuildingLevels(catalog, castleLevel, savedLevels = {}) {
  const levels = minimumBuildingLevels(catalog, castleLevel);
  for (const [buildingId, rawValue] of Object.entries(savedLevels || {})) {
    const building = catalog.buildings.get(buildingId);
    if (!building) continue;
    levels[buildingId] = Math.max(Number(levels[buildingId] || 0), Math.min(building.maxLevel, Math.max(0, Math.trunc(Number(rawValue) || 0))));
  }
  levels.castle = Math.max(Number(levels.castle || 0), Number(castleLevel) || 1);
  return levels;
}

function adjustedConstructionTime(baseSeconds, settings) {
  const speed = Math.max(0, Number(settings.constructionSpeedPercent) || 0)
    + Math.max(0, Number(settings.constructionSpeedBoostPercent) || 0);
  let remaining = Math.ceil(Math.max(0, Number(baseSeconds) || 0) / (1 + speed / 100));
  remaining = Math.max(0, remaining - freeSecondsForVip(settings.vipLevel));
  for (let count = 0; count < guildHelpCount(settings) && remaining > 0; count += 1) {
    remaining = Math.max(0, Math.ceil(remaining - Math.max(60, remaining * 0.01)));
  }
  return remaining;
}

export function createCastlePlan(catalog, state, targetCastleLevel, targetManaStage = state.settings.castleTargetManaStage, options = {}) {
  const castle = catalog.buildings.get("castle");
  const current = Math.min(castle.maxLevel, Math.max(1, Math.trunc(Number(state.settings.castleLevel) || 1)));
  const targetBuildingId = options.targetBuildingId || "castle";
  const selectedBuilding = catalog.buildings.get(targetBuildingId);
  if (!selectedBuilding) throw new Error(`不明な施設: ${targetBuildingId}`);
  const target = targetBuildingId === "castle"
    ? Math.min(castle.maxLevel, Math.max(current, Math.trunc(Number(targetCastleLevel) || current)))
    : current;
  const currentManaStage = current >= castle.maxLevel
    ? Math.min(catalog.maxManaStage, Math.max(0, Math.trunc(Number(state.settings.castleManaStage) || 0)))
    : 0;
  const normalizedTargetManaStage = targetBuildingId === "castle" && target >= castle.maxLevel
    ? Math.min(catalog.maxManaStage, Math.max(currentManaStage, Math.trunc(Number(targetManaStage) || 0)))
    : 0;
  const effectiveLevels = effectiveBuildingLevels(catalog, current, state.buildingLevels);
  const steps = [];
  const completed = new Set();
  const visiting = new Set();
  const issues = [];
  const selectedTarget = targetBuildingId === "castle"
    ? target
    : Math.min(selectedBuilding.maxLevel, Math.max(Number(effectiveLevels[targetBuildingId] || 0), Math.trunc(Number(options.targetBuildingLevel) || 0)));
  const addBuilding = (buildingId, targetLevel) => {
    const building = catalog.buildings.get(buildingId);
    if (!building) { issues.push(`不明な施設: ${buildingId}`); return; }
    const normalized = Math.min(building.maxLevel, Math.max(0, Math.trunc(Number(targetLevel) || 0)));
    for (let level = Number(effectiveLevels[buildingId] || 0) + 1; level <= normalized; level += 1) {
      const key = `${buildingId}\0${level}`;
      if (completed.has(key)) continue;
      if (visiting.has(key)) throw new Error(`施設の前提条件が循環しています: ${buildingId}:${level}`);
      visiting.add(key);
      const data = building.levels.get(level);
      if (!data) { issues.push(`施設データ未収録: ${buildingId}:${level}`); visiting.delete(key); continue; }
      for (const requirement of data.requirements) addBuilding(requirement.buildingId, requirement.level);
      steps.push({ buildingId, level, baseSeconds: data.baseTimeSeconds, adjustedSeconds: adjustedConstructionTime(data.baseTimeSeconds, state.settings), costs: { ...data.costs } });
      completed.add(key);
      visiting.delete(key);
    }
  };
  addBuilding(targetBuildingId, selectedTarget);
  for (let stage = currentManaStage + 1; stage <= normalizedTargetManaStage; stage += 1) {
    const data = catalog.manaStages.get(stage);
    if (!data) { issues.push(`城マナ強化データ未収録: ${stage}`); continue; }
    steps.push({
      buildingId: "castle",
      level: castle.maxLevel,
      manaStage: stage,
      baseSeconds: data.baseTimeSeconds,
      adjustedSeconds: adjustedConstructionTime(data.baseTimeSeconds, state.settings),
      costs: { ...data.costs },
    });
  }
  const grouped = new Map();
  const totalCosts = Object.fromEntries(CASTLE_RESOURCE_KEYS.map((key) => [key, 0]));
  for (const step of steps) {
    if (!grouped.has(step.buildingId)) grouped.set(step.buildingId, { buildingId: step.buildingId, currentLevel: Number(effectiveLevels[step.buildingId] || 0), targetLevel: Number(effectiveLevels[step.buildingId] || 0), baseSeconds: 0, adjustedSeconds: 0, costs: Object.fromEntries(CASTLE_RESOURCE_KEYS.map((key) => [key, 0])) });
    const row = grouped.get(step.buildingId);
    row.targetLevel = Math.max(row.targetLevel, step.level);
    row.baseSeconds += step.baseSeconds;
    row.adjustedSeconds += step.adjustedSeconds;
    for (const key of CASTLE_RESOURCE_KEYS) { row.costs[key] += Number(step.costs[key] || 0); totalCosts[key] += Number(step.costs[key] || 0); }
  }
  const gemCosts = gemCostsFor(catalog, totalCosts, state.settings.resources || {});
  return {
    currentCastleLevel: current,
    targetCastleLevel: target,
    currentManaStage,
    targetManaStage: normalizedTargetManaStage,
    targetBuildingId,
    targetBuildingLevel: selectedTarget,
    effectiveLevels,
    steps,
    buildings: [...grouped.values()],
    totals: { baseSeconds: steps.reduce((sum, step) => sum + step.baseSeconds, 0), adjustedSeconds: steps.reduce((sum, step) => sum + step.adjustedSeconds, 0), costs: totalCosts, gemCosts, totalGems: Object.values(gemCosts).reduce((sum, value) => sum + Number(value || 0), 0) },
    issues: [...new Set(issues)],
  };
}

export function buildingLevelsAfterCastleStep(plan, selectedStep, currentCastleLevel, currentManaStage = 0, currentBuildingLevels = {}) {
  const selectedIndex = plan.steps.findIndex(
    (step) => step.buildingId === selectedStep.buildingId
      && step.level === selectedStep.level
      && Number(step.manaStage || 0) === Number(selectedStep.manaStage || 0),
  );
  if (selectedIndex < 0) {
    return {
      castleLevel: Math.max(1, Number(currentCastleLevel) || 1),
      castleManaStage: Math.max(0, Number(currentManaStage) || 0),
      buildingLevels: { ...currentBuildingLevels },
    };
  }
  let castleLevel = Math.max(1, Number(currentCastleLevel) || 1);
  let castleManaStage = castleLevel >= 25 ? Math.max(0, Number(currentManaStage) || 0) : 0;
  const buildingLevels = { ...currentBuildingLevels };
  for (const completed of plan.steps.slice(0, selectedIndex + 1)) {
    if (completed.buildingId === "castle") {
      castleLevel = Math.max(castleLevel, completed.level);
      castleManaStage = Math.max(castleManaStage, Number(completed.manaStage || 0));
    } else {
      buildingLevels[completed.buildingId] = Math.max(
        Number(buildingLevels[completed.buildingId] || 0),
        completed.level,
      );
    }
  }
  return { castleLevel, castleManaStage, buildingLevels };
}

export function castleProgressLabel(level, manaStage = 0) {
  const suffix = Number(level) >= 25 && Number(manaStage) > 0 ? `-${Number(manaStage)}` : "";
  return `Lv.${Number(level)}${suffix}`;
}
